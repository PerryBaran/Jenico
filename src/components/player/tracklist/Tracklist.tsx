import React, { useState, useEffect, Dispatch, SetStateAction } from 'react';
import style from './tracklist.module.css';
import { menu } from '../../../media/icons/index';
import { SongInfo, Songs } from '../../../Interface';

function Tracklist(props: {data: SongInfo[], setSongIndex: Dispatch<SetStateAction<number>>, albumIndex: number, setAlbumIndex: Dispatch<SetStateAction<number>>}) {
    const {data, setSongIndex, albumIndex, setAlbumIndex} = props;
    const [pageIndex, setPageIndex] = useState(albumIndex);

    useEffect(() => {
        setPageIndex(albumIndex)
    }, [albumIndex]);

    const scrollTrackList = (forwards = true) => {
        const playlistLength = data.length;
        if (forwards) {
            const tempPageIndex = pageIndex + 1
            if (tempPageIndex >= playlistLength) {
                setPageIndex(0);
            } else {
                setPageIndex(tempPageIndex);
            }
        } else {
            const tempPageIndex = pageIndex - 1
            if (tempPageIndex < 0) {
                setPageIndex(playlistLength - 1);
            } else {
                setPageIndex(tempPageIndex);
            }
        }
    };

    const setSongTracklist = (index: number) => {
        setAlbumIndex(pageIndex);
        setSongIndex(index);
    };

    return (
        <div className={style.tracklistContainer}>
            <div className={style.tracklist}>
                <img src={data[pageIndex]?.art} alt='cover art'/>
                <div>
                    <button onClick={() => scrollTrackList(false)}>&lt;</button>
                    <h4>{data[pageIndex]?.title}</h4>
                    <button onClick={() => scrollTrackList()}>&gt;</button>
                </div>
                <ul>
                    {data[pageIndex]?.songs.map((song: Songs) => {
                        const index = data[pageIndex]?.songs.indexOf(song);
                        return (
                            <li key={song.name}>
                                <button className={style.tracklistButton} onClick={() => setSongTracklist(index)}>{song.name}</button>
                            </li>
                        )
                    })}
                </ul>
            </div>
            <button><img src={menu} alt='tracklist'/></button>
        </div>
    );
}

export default Tracklist;