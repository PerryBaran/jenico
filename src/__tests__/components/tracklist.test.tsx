import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Tracklist from "../../components/player/tracklist/Tracklist";
import mockData from "../../media/songInfo";

const ALBUM_INDEX = 1;

describe("Tracklist", () => {
  const validProps = {
    data: mockData,
    selectedSong: mockData[ALBUM_INDEX].songs[0].name,
    selectedAlbum: mockData[ALBUM_INDEX],
    handleSelectedSong: jest.fn(),
    handlePlaying: jest.fn(),
  };

  afterEach(cleanup);

  test("snapshot", () => {
    const { asFragment } = render(
      <Tracklist
        data={validProps.data}
        selectedSong={validProps.selectedSong}
        selectedAlbum={validProps.selectedAlbum}
        handleSelectedSong={validProps.handleSelectedSong}
        handlePlaying={validProps.handlePlaying}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("content", () => {
    render(
      <Tracklist
        data={validProps.data}
        selectedSong={validProps.selectedSong}
        selectedAlbum={validProps.selectedAlbum}
        handleSelectedSong={validProps.handleSelectedSong}
        handlePlaying={validProps.handlePlaying}
      />
    );

    expect(
      screen.getByAltText(`${validProps.selectedAlbum.title} cover art`)
    ).toHaveAttribute("src", validProps.selectedAlbum.art);
    expect(screen.getByText("<")).toBeInstanceOf(HTMLButtonElement);
    expect(screen.getByText(validProps.selectedAlbum.title)).toBeInstanceOf(
      HTMLHeadingElement
    );
    expect(screen.getByText(">")).toBeInstanceOf(HTMLButtonElement);
    expect(screen.getByAltText("tracklist")).toHaveAttribute("src", "menu.png");

    validProps.selectedAlbum.songs.forEach((song, i) => {
      expect(screen.getByText(song.name)).toBeInstanceOf(HTMLButtonElement);
    });
  });

  test("scrollTrackList", () => {
    render(
      <Tracklist
        data={validProps.data}
        selectedSong={validProps.selectedSong}
        selectedAlbum={validProps.selectedAlbum}
        handleSelectedSong={validProps.handleSelectedSong}
        handlePlaying={validProps.handlePlaying}
      />
    );

    expect(
      screen.queryByText(validProps.data[ALBUM_INDEX + 1].title)
    ).not.toBeInTheDocument();
    fireEvent.click(screen.getByText(">"));
    expect(
      screen.getAllByText(validProps.data[ALBUM_INDEX + 1].title)[0]
    ).toBeInTheDocument();
    expect(
      screen.queryByText(validProps.selectedAlbum.title)
    ).not.toBeInTheDocument();
    fireEvent.click(screen.getByText("<"));
    expect(
      screen.queryByText(validProps.data[ALBUM_INDEX + 1].title)
    ).not.toBeInTheDocument();
    expect(
      screen.getByText(validProps.selectedAlbum.title)
    ).toBeInTheDocument();
  });

  test("handleSelectSong and handlePlaying", () => {
    render(
      <Tracklist
        data={validProps.data}
        selectedSong={validProps.selectedSong}
        selectedAlbum={validProps.selectedAlbum}
        handleSelectedSong={validProps.handleSelectedSong}
        handlePlaying={validProps.handlePlaying}
      />
    );
    const buttonName = validProps.data[ALBUM_INDEX].songs[3].name;

    expect(validProps.handleSelectedSong).toBeCalledTimes(0);
    expect(validProps.handlePlaying).toBeCalledTimes(0);
    fireEvent.click(screen.getByText(buttonName));
    expect(validProps.handleSelectedSong).toBeCalledTimes(1);
    expect(validProps.handleSelectedSong).toBeCalledWith(buttonName);
    expect(validProps.handlePlaying).toBeCalledTimes(1);
    expect(validProps.handlePlaying).toBeCalledWith(true);
  });
});
