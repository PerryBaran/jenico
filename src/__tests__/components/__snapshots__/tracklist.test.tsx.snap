// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tracklist snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="tracklist"
    >
      <img
        alt="Dreaming of Detuned Love cover art"
        src="DreamingOfDetunedLove.png"
      />
      <div>
        <button>
          &lt;
        </button>
        <h4>
          Dreaming of Detuned Love
        </h4>
        <button>
          &gt;
        </button>
      </div>
      <ul>
        <li
          class="highlight"
        >
          <button>
            Slumber
          </button>
        </li>
        <li>
          <button>
            Your Careless Embrace
          </button>
        </li>
        <li>
          <button>
            Restless Thoughts
          </button>
        </li>
        <li>
          <button>
            Detuned Love
          </button>
        </li>
        <li>
          <button>
            Falling into the Void
          </button>
        </li>
      </ul>
    </div>
    <img
      alt="tracklist"
      class="tracklistMenu"
      src="menu.png"
    />
  </div>
</DocumentFragment>
`;
